rfc: trusted k8s images (#2648)
edgelesssys/constellation: rfc/XXX-trusted-kubernetes-images.md

# RFC XXX: Trusted Kubernetes Images

Kubernetes control plane images should be verified by the Constellation installation.

## The Problem

When we bootstrap the Constellation Kubernetes cluster, `kubeadm` places a set
of static pods for the control plane components into the filesystem. The
manifests refer to images in a registry beyond the users' control, and the
image content is not reproducible.

This is obviously a trust issue, because the Kubernetes control plane is
part of Constellation's TCB, but it is also a problem when Constellation is set
up in a restricted environment where this repo is not available.

## Requirements

1. In a default installation, Constellation must verify Kubernetes control plane images.
2. Users must be able to override the image repository for Kubernetes control plane images.

> 3. Users must be able to override the verification mechanism for Kubernetes control plane images.

Author: Reduced the scope accordingly.

registry.k8s.io and ship hashes with the CLI. Users can override the image
per control plane component.

### Image Hashes

We are concerned with the following control plane images (for v1.27.7):

- registry.k8s.io/kube-apiserver:v1.27.7
- registry.k8s.io/kube-controller-manager:v1.27.7
- registry.k8s.io/kube-scheduler:v1.27.7
- registry.k8s.io/coredns/coredns:v1.10.1
- registry.k8s.io/etcd:3.5.9-0

When a new Kubernetes version is added to `/internal/versions/versions.go`, we
generate the corresponding list of images with `kubeadm config images list` and
probe their hashes on registry.k8s.io. Generating the list of images this way
must happen offline to prevent `kubeadm` from being clever. These hashes are
added to `versions.go` as a mapping from component to pinned image:

```golang
V1_27: {
  ClusterVersion: "v1.27.7",
  // ...
  ControlPlaneImages: {
    "kube-apiserver": "registry.k8s.io/kube-apiserver:v1.27.7@sha256:<...>", 
    "kube-controller-manager": "registry.k8s.io/kube-controller-manager:v1.27.7@sha256:<...>", 
    "kube-scheduler": "registry.k8s.io/kube-scheduler:v1.27.7@sha256:<...>", 
    "etcd": "registry.k8s.io/etcd:3.5.9-0@sha256:<...>",

>     "coredns": "registry.k8s.io/coredns/coredns:v1.10.1@sha256:<...>",

Author: Changed to `KubernetesImages`.

  },
}
```

### Configuration Options

Users can override the images in their `constellation-conf.yaml`, using the
same nested structure as `versions.go` (*to be elaborated on*).

### Cluster Init

During cluster initialization, we need to tell `kubeadm` that we want to use
the embedded image references instead of the default ones. For that, we
populate the
[`InitConfiguration.Patches`](https://pkg.go.dev/k8s.io/kubernetes@v1.27.7/cmd/kubeadm/app/apis/kubeadm/v1beta3#InitConfiguration)
with a list of patch files that replace the container image with the pinned
alternative.

This method does not work for coredns, which can only be customized with the
[`ClusterConfiguration.DNS`](https://pkg.go.dev/k8s.io/kubernetes@v1.27.7/cmd/kubeadm/app/apis/kubeadm/v1beta3#ClusterConfiguration)

> section. It's a bit of a hack, but we can parse the registry part and the

Author: The hack part is that it would look like you could do

```yaml
"coredns": "my.custom.registry/my-custom-coredns:123@sha256:..."
```

but this can't be modeled with `ClusterConfiguration.DNS` because the image path is fixed.

In the spirit of https://github.com/edgelesssys/constellation/pull/2648/files#r1406263649 it would make sense to manage coredns as a microservice.

> section. It's a bit of a hack, but we can parse the registry part and the

Author: Adjusted wording.

tag/hash part from the full image and set it in that `DNS` section of the
config.

The patches need to be materialized to the stateful filesystem by the
bootstrapper. As there are a lot of similarities between these patches and
`components.Components`, we could extend that definition to also allow inline

> file content as an alternative to remote URI and hash. Alternatively, we could

Author: Implemented.

move the creation of the kubeadm configuration to the cli and let the
bootstrapper only override fields that are discovered in the cluster, like

> cloud provider specifics.

Author: I guess this idea is orthogonal to the proposal, but might make this and future customizations of kubeadm easier to do. Agreed that this is a larger shift, and happy to drop it from here.

> cloud provider specifics.

Author: Removed from the proposal.

# RFC XXX: Trusted Kubernetes Images

Kubernetes control plane images should be verified by the Constellation installation.

## The Problem

When we bootstrap the Constellation Kubernetes cluster, `kubeadm` places a set
of static pods for the control plane components into the filesystem. The
manifests refer to images in a registry beyond the users' control, and the

> image content is not reproducible.

Reviewer: ```suggestion
container image content is not reproducible.
```
nit

This is obviously a trust issue, because the Kubernetes control plane is
part of Constellation's TCB, but it is also a problem when Constellation is set
up in a restricted environment where this repo is not available.

## Requirements

1. In a default installation, Constellation must verify Kubernetes control plane images.

Out of scope:

- Customization of the image repository or image content.
  - This is orthogonal to image verification and will be subject of a separate
    RFC.
- Reproducibility from github.com/kubernetes/kubernetes to registry.k8s.io and
  the associated chain of trust.
  - It is not clear whether Kubernetes images can be reproduced at all [1].
    Either way, the likely threat is a machine-in-the-middle attack between
    Constellation and registry.k8s.io. A desirable addition to this proposal
    could be verification of image signatures [2].
- Container registry trust & CA certificates.
  - This is also orthogonal to image verification.

[1]: https://github.com/kubernetes/kubernetes/blob/master/build/README.md#reproducibility
[2]: https://kubernetes.io/docs/tasks/administer-cluster/verify-signed-artifacts/

## Solution

> Kubernetes control plane images are going to be pinned by hash and verified by

Reviewer: ```suggestion
Kubernetes control plane images are going to be pinned by a hash, which is verified by
```
nit: reword due to ambiguity with the above mentioned signature verification of container images

> the CRI. Images hashes are added to the Constellation codebase when support for

Author: ```suggestion
the CRI. Image hashes are added to the Constellation codebase when support for
```

must happen offline to prevent `kubeadm` from being clever. These hashes are
added to `versions.go` as a mapping from component to pinned image:

```golang
V1_27: {
  ClusterVersion: "v1.27.7",
  // ...
  KubernetesImages: {
    "kube-apiserver": "registry.k8s.io/kube-apiserver:v1.27.7@sha256:<...>", 
    "kube-controller-manager": "registry.k8s.io/kube-controller-manager:v1.27.7@sha256:<...>", 
    "kube-scheduler": "registry.k8s.io/kube-scheduler:v1.27.7@sha256:<...>", 
    "etcd": "registry.k8s.io/etcd:3.5.9-0@sha256:<...>",
    "coredns": "registry.k8s.io/coredns/coredns:v1.10.1@sha256:<...>",
  },
}
```

### Cluster Init

During cluster initialization, we need to tell `kubeadm` that we want to use
the embedded image references instead of the default ones. For that, we
populate the
[`InitConfiguration.Patches`](https://pkg.go.dev/k8s.io/kubernetes@v1.27.7/cmd/kubeadm/app/apis/kubeadm/v1beta3#InitConfiguration)
with a list of [JSON Patch](https://datatracker.ietf.org/doc/html/rfc6902)
files that replace the container image with the pinned alternative.

The patches need to be written to the stateful filesystem by the

> bootstrapper. This is very similar to what `components.Component`, which also

Author: ```suggestion
bootstrapper. This is very similar to `components.Component`, which also
```

place Kubernetes-related data onto the filesystem:

```go
type Component struct {
  URL         string
  Hash        string
  InstallPath string
  Extract     bool
}
```

Components are handled by the installer, where the convention currently expects
HTTP URLs that are to be downloaded. We can extend this by allowing other forms
of URI schemes:
[data URLs](https://developer.mozilla.org/en-US/docs/web/http/basics_of_http/data_urls).
A patch definition as Component would look like this:

```go
patch := &components.Component{
  URL: "data:application/json;base64,W3sib3AiOiJyZXBsYWNlIiwicGF0aCI6Ii9zcGVjL2NvbnRhaW5lcnMvMC9pbWFnZSIsInZhbHVlIjoicmVnaXN0cnkuazhzLmlvL215LWNvbnRyb2wtcGxhbmUtaW1hZ2U6djEuMjcuN0BzaGEyNTY6Li4uIn1dCg=="
  InstallPath: "/opt/kubernetes/patches/kube-apiserver+json.json"

>   // InstallPath and Extract deliberately left empty.

Reviewer 2: ```suggestion
  // Hash and Extract deliberately left empty.
```

Author: Good catch, thanks!
